### ISSUE TITLE
params used in config file set to [:] if not defined outside main.nf

### ISSUE THREAD

**Issue by MaxUlysse** [nextflow-io/nextflow#486]
I wanted to use a `params` for specifying a `tag` when defining a container (like `debian:${params.tag}`).
When I define `tag` in the `config` file or via the command line, it works well.
But if I define `tag` only in the `main.nf` script, there is some erratic behaviors.

Here is a minimal example:

`nextflow.config`:
```
params.tag = '8.9'
docker.enabled = true
process.$test.container = "debian:${params.tag}"
```

`main.nf`:
```groovy
#!/usr/bin/env nextflow
params.tag = '8.6' // Define default tag
println "params.tag: $params.tag"
process test {
  tag {workflow.container.$test}
  input:
  output:
  script:
    """
    touch test
    """
}
```

All works well:
```bash
≻nextflow run .
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [voluptuous_raman] - revision: 4684eeda31
params.tag: 8.9
[warm up] executor > local
[79/ea9b32] Submitted process > test (debian:8.9)

nextflow run . --tag 8
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [modest_panini] - revision: 4684eeda31
params.tag: 8
[warm up] executor > local
[cb/ef58e2] Submitted process > test (debian:8)
```

Removing `params.tag` definition from config file
new `nextflow.config`:
```
docker.enabled = true
process.container = "debian:${params.tag}"
```
Normal behavior
```bash
>nextflow run . --tag 8
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [peaceful_roentgen] - revision: 4684eeda31
params.tag: 8
[warm up] executor > local
[53/b0a080] Submitted process > test (debian:8)
```

Expected behavior
```bash
>nextflow run .
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [cheesy_agnesi] - revision: 8e097470c7
params.tag: 8.6
[warm up] executor > local
[88/fda415] Submitted process > test (debian:8.6)
```

What I really got:
```bash
>nextflow run .
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [jovial_bohr] - revision: 4684eeda31
params.tag: [:]
[warm up] executor > local
[cd/a92ac0] Submitted process > test (debian:[:])
ERROR ~ Error executing process > 'test (debian:[:])'

Caused by:
  Process `test (debian:[:])` terminated with an error exit status (125)

Command executed:

  touch test

Command exit status:
  125

Command output:
  (empty)

Command error:
  docker: invalid reference format.
  See 'docker run --help'.

Work dir:
  /home/max/workspace/github/TAG/work/cd/a92ac0b934e212c380e377d0f058a0

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`

 -- Check '.nextflow.log' file for details
```

Trying to get the container definition inside the `main.nf` script:
new `nextflow.config`:
```
docker.enabled = true
```
new `main.nf`:
```groovy
#!/usr/bin/env nextflow
params.tag = '8.6' // Define default tag
println "params.tag: $params.tag"
process test {
  container "debian:${params.tag}"
  tag {workflow.container.$test}
  input:
  output:
  script:
    """
    touch test
    """
}
```

Working, but I honestly don't know where the container name is stored:

```bash
╰─≻nextflow run .
N E X T F L O W  ~  version 0.25.6
Launching `./main.nf` [kickass_knuth] - revision: bd08825bd7
params.tag: 8.6
[warm up] executor > local
[94/05e9b9] Submitted process > test (null)

# I got the container name wrong, but the container used is the good one
≻tail work/94/05e9b905017394d63af04a7d5f6bd4/.command.run
tee .command.out < "$COUT" &
tee1=$!
tee .command.err < "$CERR" >&2 &
tee2=$!
(
docker run -i -v /home/max/workspace/github/TAG/work/94/05e9b905017394d63af04a7d5f6bd4:/home/max/workspace/github/TAG/work/94/05e9b905017394d63af04a7d5f6bd4 -v "$PWD":"$PWD" -w "$PWD" --entrypoint /bin/bash --name $NXF_BOXID debian:8.6 -c "/bin/bash -ue /home/max/workspace/github/TAG/work/94/05e9b905017394d63af04a7d5f6bd4/.command.sh"
) >"$COUT" 2>"$CERR" &
pid=$!
wait $pid || ret=$?
wait $tee1 $tee2
```

I'm guessing that if a `params` is used in a config file, it needs to be defined in a config file as well.

**Comment by pditommaso** [comment]
Thanks for detailed report. This is a  very tricky issue to solve. The main problem when having a config file like this:

```
docker.enabled = true
process.container = "debian:${params.tag}"
```

and not `tag` option is specified on the command line, the `params.tag` is not defined when the config `process.container = "debian:${params.tag}"` is evaluated. 

Now, what happens is the config parser expand an undefined parameters to an empty object, instead of null, which makes things even more ugly. 

To summarise, I think it it should be somehow possible to avoid to get an empty map `[:]`, but in any case you will get the `process.container` attribute set to `null`.

**Comment by MaxUlysse** [comment]
No problem, we're trying to run our pipeline in production, and I was trying to use automatically containers with the tag corresponding to the version.
And since I'm lazy, I wanted to have that in only one place if possible ;-).
Finally I'm keeping the tag in the config file, not really a problem for me.
But I did think it was an issue worth reporting ;-)

**Comment by pditommaso** [comment]
Fixed by #525
